docs/design/2022-10-13-mysql-compatible-auto-increment.md

```diff
@@ -0,0 +1,110 @@
+
+In TiDB, we support generating `AUTO_INCREMENT` IDs that are **unique**, **monotone increasing**, but the IDs may **not be consecutive**.
+
+The current behaviour in TiDB is not fully-compatible with MySQL.
+
+TiDB is a distributed database, each TiDB instance caches a batch of IDs for local allocating. When it exhaust its local cached IDs, it ask for another batch. For example, TiDB instance A may get ID range [0-20000), and instance B takes away range [20000, 40000), the next batch for TiDB instance A could be [40000, 60000), that's why the IDs are not **consecutive**. The ID sequence from instance A might be ...19998, 19999, [a hole here], 40000, 40001...
+
+## Proposal
+
+This proposes that we introduce a centralized auto ID allocating service, the `AUTO_INCREMENT` ID is allocated from the centralized service and there is no caching mechanism on the TiDB layer, so the behaviour can be MySQL compatible.
+
+## Compatibility
+
+The old implementation can still be kept, its performance is better and might meet some user's needs.
+
+There are several ways to handle the compatibility.
+
+The first one is, we can rename the syntax for the old implementation, such as `AUTO_INCREMENT_FAST` or `AUTO_INCREMENT_OLD` or `DEFICIT_AUTO_INCREMENT`. The upside of this approach is that we can make the default syntax MySQL-compatible. The downside is that the new `AUTO_INCREMENT` behaviour is still a compatibility-breaker between old and new versioned TiDB.
+
+The second one is, we can keep the old things unchanged, and introduce a new syntax. The benefit of this approach is that our users will never get into trouble when they upgrade TiDB.
+
+And another way, we can reuse the `AUTO_ID_CACHE 1` table option for it:
+
+```
+CREATE TABLE t (id int key AUTO_INCREMENT) AUTO_ID_CACHE 1;
+```
+
+TiDB extends the syntax to control the AUTO ID CACHE size use the `AUTO_ID_CACHE` table option. `AUTO_ID_CACHE 1` is equivalent to no caching. Now it's of the same semantic, just that the implementation is changed. I think the this approach is the best.
+
+## Implementation
+
+### Centralized auto ID allocating service
+
+This is the core part of the implementation. There will be a single-point process, responsible for allocating auto IDs.
+
+(To be exact, the allocating service will be embeded into the TiDB process currently for simplicity, maybe in the future it can be moved to a standalone process. But wherever it is, the idea is the same)
+
+For every `AUTO_INCREMENT` ID allocation, TiDB send a request to this process. The process will `+1` internally, so the ID is **unique**, **monotone increasing**, and **consecutive**.
+
+Grpc is used for the communication protocol. This part is relatively easy so I would not go to much details in this document.
+
+```
+message AutoIDRequest {
+    int64 dbID = 1;
+    int64 tblID = 2;
+    uint64 n = 3;
```

Comment:
How can we use this for updating the service when an explicit ID has been set?
Like `INSERT INTO t SET id = 1000000, ...` which is higher than the current used auto_increment. Similar with `UPDATE t SET id = X WHERE id = Y`. To me I would assume it must be persisted as well.

Reply:
Ah, there is a Rebase message as well [here](https://github.com/pingcap/kvproto/blob/e7e21f0e9cd907d6843afbfbd375d39521113db0/proto/autoid.proto#L28). I think it should be mentioned in the design as well, since it is an important part of the auto_increment implementation.

Reply:
Yes, there is indeed a `Rebase` ... but that's too trivial ...
I think a reader would get distract by too much details, they can quickly get a big picture from the design doc, as for those low level details, the code is the source of truth.